### PR TITLE
HA Blog post

### PR DESCRIPTION
**Pull request by knolleary**
## Description

Closes https://github.com/flowforge/content/issues/96

Needs a social tile creating - will raise art request and link to the issue.

Artwork issue https://github.com/flowforge/website/issues/718

### CODE REVIEW

File: src/blog/2023/05/bringing-high-availability-to-node-red.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+layer to provide better performance and interoperability. That gets tricky when
+you want to have multiple instances sharing the same store. The context api
+doesn’t provide a way to atomically update values - so you can get into classic
+concurrency issues around two applications trying to update the same value.
+
+The other type of state is that which is implicitly maintained in a flow - even
+if the user hasn’t explicitly configured it. For example, the Smooth node can be
+used to calculate a running average value of messages passing through it. The
+node does that by keeping in memory the recent values so it can recalculate the 
+verage with each update. If you have multiple instances, then the node will be
```
Comment (MarianRaphael) [inline src/blog/2023/05/bringing-high-availability-to-node-red.md:53]:
```suggestion
average with each update. If you have multiple instances, then the node will be
```

File: src/blog/2023/05/bringing-high-availability-to-node-red.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+Docker Compose and Kubernetes. As we get into some of these HA features, we will
+need to look carefully at where we can lean on these underlying technologies -
+we don’t want to reinvent the wheel here.
+
+Our initial focus is going to be when running in a Kubernetes environment - just
+as we do with our hosted FlowForge Cloud platform. Kubernetes provides lots of
+the building blocks for creating a scalable and highly available solution, but
+it certainly doesn’t do all of the work for you.
+
+We've indentified our initial set of tasks and changes to how we'll run Node-RED
```
Comment (MarianRaphael) [inline src/blog/2023/05/bringing-high-availability-to-node-red.md:195]:
```suggestion
We've identified our initial set of tasks and changes to how we'll run Node-RED
```

**Review by MarianRaphael — APPROVED** [review]

**Review by robmarcer — APPROVED** [review]

File: src/blog/2023/05/bringing-high-availability-to-node-red.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+title: Bringing High Availability to Node-RED
+subtitle: How we are tackling the hard problems of HA in FlowForge
+description: How we are tackling the hard problems of HA in FlowForge
+date: 2023-05-08 12:00:00
+authors: ["nick-oleary"]
+# image: /images/blog/tile-chatgpt-fcn-node.jpg
+---
+
+Many companies look to deploy Node-RED into use cases that require the application
+to have a high degree of availability, reliability and scalability. Following up
```
Comment (robmarcer) [inline src/blog/2023/05/bringing-high-availability-to-node-red.md:11]:
```suggestion
to have a high degree of availability, reliability, and scalability. Following up
```

File: src/blog/2023/05/bringing-high-availability-to-node-red.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+There are two types of state to consider when thinking about a Node-RED flow:
+**explicit** and **implicit** state.
+
+Explicit state is what is programmed into the flow. For example, a flow may store
+state in Context or use an external database service. Within FlowForge we provide
+two types of context - the default in-memory context store and a database-backed 
+persistent store. Currently the database-backed store includes a memory-caching
+layer to provide better performance and interoperability. That gets tricky when
+you want to have multiple instances sharing the same store. The context api
+doesn’t provide a way to atomically update values - so you can get into classic
```
Comment (robmarcer) [inline src/blog/2023/05/bringing-high-availability-to-node-red.md:46]:
```suggestion
doesn’t provide a way to automatically update values - so you can get into classic
```

File: src/blog/2023/05/bringing-high-availability-to-node-red.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+There are two types of state to consider when thinking about a Node-RED flow:
+**explicit** and **implicit** state.
+
+Explicit state is what is programmed into the flow. For example, a flow may store
+state in Context or use an external database service. Within FlowForge we provide
+two types of context - the default in-memory context store and a database-backed 
+persistent store. Currently the database-backed store includes a memory-caching
+layer to provide better performance and interoperability. That gets tricky when
+you want to have multiple instances sharing the same store. The context api
```
Comment (robmarcer) [inline src/blog/2023/05/bringing-high-availability-to-node-red.md:45]:
```suggestion
you want to have multiple instances sharing the same store. The context API
```

File: src/blog/2023/05/bringing-high-availability-to-node-red.md [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+With our goal to minimize the Mean Time To Recovery (MTTR), we need to find a
+way to get that spare instance running as quickly as possible; if it takes just
+as long to start the spare instance as it does to restart the failed primary
+instance, then it isn’t much of an improvement.
+
+The key here is that Node-RED allows you to start the runtime without the flows
+running. That gets everything loaded and the runtime ready ahead of time. It can
+then start the flows at a moment's notice with a simple call to the runtime admin
+api.
```
Comment (robmarcer) [inline src/blog/2023/05/bringing-high-availability-to-node-red.md:121]:
```suggestion
API.
```

File: src/blog/2023/05/bringing-high-availability-to-node-red.md [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+active at the same time. This can get quite complicated to achieve safely, but
+there are a number of approaches that can be used. We’ll be exploring them as we
+continue our journey towards HA.
+
+### Editing Flows
+
+Within the Node-RED architecture, each instance also serves up its own editor.
+This is what you get when you point your web browser at it.
+
+In an HA world, once you have multiple instances running behind an HTTP load
```
Comment (robmarcer) [inline src/blog/2023/05/bringing-high-availability-to-node-red.md:147]:
```suggestion
In a HA world, once you have multiple instances running behind an HTTP load
```

File: src/blog/2023/05/bringing-high-availability-to-node-red.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+balancer, there is a tricky question of how you edit the flows. If each request
+hits a different instance, just loading the editor will result in different bits
+coming from different instances. That can typically be solved at the load balancer
+level by creating sticky-sessions; ensuring for a given client, each request is
+routed to a consistent instance. That solves part of the issue, but the next
+challenge is what to do when the Deploy button is pressed. That is how new flows
+are passed from the editor to the runtime. When you have multiple instances, we
+need to make sure that they all get updated. That is quite a tricky problem to
+solve with the current Node-RED APIs - and something we’ll be working both in
```
Comment (robmarcer) [inline src/blog/2023/05/bringing-high-availability-to-node-red.md:156]:
```suggestion
solve with the current Node-RED APIs - and something we’ll be working on both in
```

**Review by robmarcer — CHANGES_REQUESTED** [review]

File: src/blog/2023/05/bringing-high-availability-to-node-red.md (hunk 4 above)
Comment (knolleary) [reply src/blog/2023/05/bringing-high-availability-to-node-red.md:46]: No - atomically is the right word.

**Review by robmarcer — APPROVED** [review]

File: src/blog/2023/05/bringing-high-availability-to-node-red.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+subtitle: How we are tackling the hard problems of HA in FlowForge
+description: How we are tackling the hard problems of HA in FlowForge
+date: 2023-05-08 12:00:00
+authors: ["nick-oleary"]
+# image: /images/blog/tile-chatgpt-fcn-node.jpg
+---
+
+Many companies look to deploy Node-RED into use cases that require the application
+to have a high degree of availability, reliability, and scalability. Following up
+our [previous post on the subject](../02/highly-available-node-red.md), in this
```
Comment (robmarcer) [inline src/blog/2023/05/bringing-high-availability-to-node-red.md:12]:
```suggestion
our [previous post on the subject](/blog/02/highly-available-node-red), in this
```

**Comment by iskerrett** [comment]
This post seems ready for publishing.  @knolleary do you agree?  @ZJvandeWeg you are the last reviewer

**Review by MarianRaphael — CHANGES_REQUESTED** [review]

**Comment by ZJvandeWeg** [comment]
I'll first let @knolleary decide if this is indeed words he's comfortable with, but I think this blog would help and I'm supportive of merging sooner rather than later

**Review by MarianRaphael — APPROVED** [review]

File: src/blog/2023/05/bringing-high-availability-to-node-red.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
+---
+title: Bringing High Availability to Node-RED
+subtitle: How we are tackling the hard problems of HA in FlowForge
+description: How we are tackling the hard problems of HA in FlowForge
+date: 2023-05-08 12:00:00
```
Comment (knolleary) [inline src/blog/2023/05/bringing-high-availability-to-node-red.md:5]:
```suggestion
date: 2023-06-02 12:00:00
```

File: src/blog/2023/05/bringing-high-availability-to-node-red.md [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+authors: ["nick-oleary"]
+image: /images/blog/tile-ha-nr.jpg
+---
+
+Many companies look to deploy Node-RED into use cases that require the application
+to have a high degree of availability, reliability, and scalability. Following up
+our [previous post on the subject](/blog/02/highly-available-node-red), in this
```
Comment (knolleary) [inline src/blog/2023/05/bringing-high-availability-to-node-red.md:12]:
```suggestion
our [previous post on the subject](/blog/2023/02/highly-available-node-red/), in this
```

**Pull request merged**